source/fundamentals/bson.txt

```diff
@@ -0,0 +1,231 @@
+   :class: singlecol
+
+Overview
+--------
+
+In this guide, you can learn about how the Go Driver handles conversions
+between BSON and Go objects. The process of converting a Go object to
+BSON is called **marshalling**, while the reverse process is called **unmarshalling**. 
+
+You should read this guide if you want to learn more about how the Go Driver
+represents BSON data or need to adjust default marshalling
+and unmarshalling behaviors.
+
+Data Types
+----------
+
+MongoDB stores documents in a binary representation called :manual:`BSON
+</reference/bson-types/>` that allows for easy and flexible data processing. The Go
+Driver represents BSON data using two families of types: ``D`` and ``Raw``.
+
+The ``D`` family contains four types:
+
+- ``D``: An ordered representation of a BSON document (slice)
+- ``M``: An unordered representation of a BSON document (map)
+- ``A``: An ordered representation of a BSON array
+- ``E``: A single element inside a D type
+
+Of these, you can use the ``D`` and ``M`` types to
+build representations of BSON using native Go types. 
+
+The following example demonstrates how to construct a query filter using the
+``bson.D`` type to match documents with a ``quantity`` field value greater
+than 100:
+
+.. code-block:: go
+   
+   filter := bson.D{{"quantity", bson.D{{"$gt", 100}}}}
+
+You can use ``Raw`` types to validate and retrieve elements from a slice
+of bytes, which is useful if you don't want to unmarshal BSON into
+another type. 
+
+In Go, a **struct** is a collection of data fields with declared data
+types. The Go Driver can marshal/unmarshal structs and other native Go
+types to/from BSON using a `configurable codec system <https://pkg.go.dev/go.mongodb.org/mongo-driver/bson/bsoncodec>`_.
+
+For more information on how the Go Driver handles BSON data, see the
+`bson package API documentation
+<https://pkg.go.dev/go.mongodb.org/mongo-driver/bson>`_.
+
+Struct Tags
+-----------
+
+You can modify the default marshalling behavior of the Go Driver using
+**struct tags**, which are optional pieces of metadata attached to
+struct fields. The most common use of struct tags is for specifying the
+field name in the BSON document that corresponds to the struct field.
+The following table describes the additional struct tags that you can
+use with the Go Driver:
+
+.. list-table::
+   :widths: 25 75
+   :header-rows: 1
+
+   * - Struct Tag
+     - Description
+
+   * - ``omitempty``
+     - The field will not be marshalled if it is set to the zero value
+       corresponding to the field type.
+
+   * - ``minsize``
+     - If the field type is type int64, uint, uint32, or uint64 and the value of
+       the field can fit in a signed int32, the field will be serialized
+       as a BSON int32 rather than a BSON int64.
+
+   * - ``truncate``
+     - If the field type is a non-float numeric type, BSON doubles
+       unmarshalled into that field will be trucated at the decimal point.
+
+   * - ``inline``
```

Comment:
Struct flattening can be a bit confusing sometimes for users so I think it'd be useful to have a concrete example of how values will roundtrip from struct to BSON back to struct when using this tag -- the serde documentation has a nice page about this: https://serde.rs/attr-flatten.html

Reply:
added this to the example